[6.18] realtek: various multicast snooping fixes and improvements

T-X opened the pull request:
The following patchset is to get the rtl83xx multicast snooping behaviour closer to the one of a pure software Linux bridge in the following ways and to fix a few issues:

Firstly, a plain Linux bridge would avoid applying multicast snooping if either no IGMP/MLD querier is present or if multicast snooping is disabled. This is to avoid potential multicast packet loss, IGMP/MLD queries are crucial for a proper multicast snooping behaviour. Patch 8+9/11 tries to achieve this, by adding a DSA/switchdev notifier to let the Linux bridge inform rtl83xx if multicast snooping is fully active.

Secondly, a plain Linux bridge would avoid any forwarding if for a snoopable multicast destination no MLD reports were heard. rtl83xx on the other hand however floods a multicast packets to all ports if it has not detected any matching multicast listener. Patch 10/11 tries to adopt the behaviour from the plain Linux bridge, to cut unneccessary multicast packet transmissions.

Thirdly, multicast packets need to always be forwarded to multicast router ports, as required by RFC4541, section 2.1.2.1. This is not only to allow the usage layer 3 multicast routers, but also to avoid multicast (and IPv6 unicast) packet loss in certain scenarios. Patch 11/11 implements this forwarding to multicast router ports behaviour.

Patches 1 to 7 are some cleanup work/preparations, like using more macros to enhance readability and adding new debugfs files. To hopefully make it easier to verify the altered behaviour.

Tested-on: ZyXEL GS1900-24HP v1

---

Rebased on top of Linux v6.18 / PR #21181, commit 43f07c23c35a.

Roadmap to remove the "Draft" status + merge this:

- [ ]: make a new PR with the first few patches, the cleanup and refactoring ones, from this PR, get them merged in OpenWrt
- [ ]: Linux bridge: get mc_active merged in the upstream Linux kernel
- [ ]: Linux bridge/DSA/switchdev: get DSA port_mdb_set_mrouter() handler merged in the upstream Linux kernel, CC openwrt-devel
- [ ]: have Linux kernel bump to 6.18 (#21181) merged in OpenWrt

After the above:

- [ ]: remove "Draft" status from this PR
- [ ]: merge this PR

T-X committed:
realtek: replace additional "magic numbers" with macros

Increase readability by replacing more numbers and bit operations
with macros in order to give them distinct, expressive names.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: add portmasks, vlan_profiles and vlans to debugfs tables

Prepare portmasks, vlan profiles and vlans for changes
in upcoming patches by exporting them to debugfs for easier
debugging and monitoring.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: setup special portmasks only once

The special all-ports multicast portmasks table entry would be written
twice, redundantly as vlan_profile_setup() is called twice. Which is
unnecessary. Writing this reserved multicast portmask entry is
independent of a VLAN profile, therefore let's move this to a more
suitable position.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: split MDB to L2 hash/cam into smaller functions

Restructure the MDB to L2 hash/cam addition and removal into smaller
functions in order to make them easier to follow and allow reusal
by follow-up patches.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: only apply multicast snooping with an active querier

Currently, when enabling multicast snooping but forgetting to enable a
multicast querier on the link, then this will lead to multicast packet
loss (plus IPv6 unicast packet loss, as it depends on multicast). Or
if the multicast querier was configured on another host, but this host
crashes for example. MLD queries are needed to reliably and periodically
get MLD reports from multicast listeners.

The Linux bridge already tracks if it can safely and actively use
multicast snooping for that, that is if both multicast snooping is
enabled and a multicast querier is present for a particular VID and
protocol family. With this patch we hook into the corresponding, new DSA
notifier to adopt this state and safety measure from the Linux bridge.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: avoid flooding unknown multicast groups

The goal of IGMP/MLD snooping is to save multicast transmissions, to
forward a multicast packet with a snoopable multicast destination only
to ports where a host has actively signalled interest in it through MLD
reports.

However in the special case of no interested multicast listener we would
so far flood such a multicast packet instead. Which is contrary to our
original goal. This patch changes this by avoiding forwarding snoopable
multicast packets with no listeners altogether.

We need to make sure that on the other hand we still keep flooding
multicast packets with an unsnoopable multicast destination, as stated
by RFC4541, section 2.1.2.2 (224.0.0.x) + section 3 (ff02::1).
Therefore when disabling the VLAN L2 unknown multicast flooding setting
we are adding virtual entries for these unsnoopable addresses to ensure
we keep flooding these.

To not flood snoopable but unknown multicast addresses is also the default
behaviour of a pure software Linux bridge.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X commented:
@svanheule @logicog @plappermaul @janh

T-X committed:
realtek: forward multicast packets to multicast router ports

With this patch a multicast packet is not only forwarded to detected
multicast listeners, but also to detected multicast routers or manually
configured multicast router ports.

For one thing, this now allows using layer 3 multicast routers behind a
switch port.

Even more importantly though this forwarding is required by
RFC4541, section 2.1.2.1. Otherwise not only for routable multicast
packets but also for link-local multicast packet loss might occur
if the elected multicast querier is not on this device.
And hence IPv6 unicast would break, too, as it relies on link-local
multicast for its IPv6 Neighbor Discovery.

Fixes: cde31976e375 ("realtek: Add support for Layer 2 Multicast")
Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

plappermaul commented:
Is this https://github.com/openwrt/openwrt/pull/18769 still needed with this patch series.

T-X commented:
> Is this #18769 still needed with this patch series.

Yes, it's still needed (and what this patchset is rebased on, it's the first, same patch in here, too).

plappermaul commented:
If this PR is superset of #18769 then you can close the other one.

This series is incredible valuable for the Realtek target. A nice balance between refactoring and new features. So I do not want to nitpick too much. Before going into details I politely ask for the following adaptions:

- Prefix the patch headers with "realtek" instead of "rtl83xx"
- Use rtl83xx prefix only for functions targeting 838x & rtl839x. For totally generic ones simply use rtl_ or rtldsa_ (a good start to invent a driver sepcific prefix).
- Use dev_dbg, dev_info, ... instead of pr_debug, ... counterparts

Thanks in advance.

T-X committed:
realtek: change function prefixes from rtl83xx_ to rtldsa_

So far the function prefix rtl83xx_ is used for functions which are
exclusive to RTL8390_FAMILY_ID but also for ones shared between
RTL8390_FAMILY_ID/RTL9300_FAMILY_ID/RTL9310_FAMILY_ID.

For a more fitting, precise use of rtl83xx_ rename the ones in dsa.c
which are also used by RTL9300_FAMILY_ID and RTL9310_FAMILY_ID
to rtldsa_.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X commented:
Changelog v3:
* all: rebased to current main branch (no changes)
* realtek: rtl83xx: fix flooding of unsnoopable multicast addresses:
  * changing title to "realtek: fix flooding of unsnoopable multicast addresses"
  * using dev_dbg() instead of pr_debug()
* realtek: change function prefixes from rtl83xx_ to rtldsa_:
  * **NEW**
* rtl83xx: replace more "magic numbers" with macros:
  * changing title to "realtek: replace more "magic numbers" with macros"
  * using dev_info() instead of pr_info()
* rtl83xx: add port_masks, vlan_profiles and vlans debugfs tables
  * changing title to "realtek: add port_masks, vlan_profiles and vlans debugfs tables"
  * using dev_dbg() instead of pr_debug()
  * moving rtl838x_vlan_profile_dump() to make it static
  * making rtl*_vlan_profile_dump() functions static and removing them from rtl83xx.h
* rtl83xx: setup special portmasks only once:
  * changing title to "realtek: setup special portmasks only once"
  * using dev_err() instead of pr_err()
* rtl83xx: split mdb to L2 hash/cam into smaller functions:
  * changing title to "realtek: split mdb to L2 hash/cam into smaller functions
  * using dev_*() instead of pr_*()
  * changing function prefixes from rtl83xx_ to rtldsa_
  * changing some common pr/dev_warn() invocations to dev_dbg()
* rtl83xx: only apply multicast snooping with an active querier
  * changing title to "realtek: only apply multicast snooping with an active querier
  * changing function prefixes from rtl83xx_ to rtldsa_
  * moving "#define RTLDSA_VLAN_PROFILE_MC_ACTIVE_V{4,6}" additions a few lines higher
* rtl83xx: avoid flooding unknown multicast groups:
  * changing title to "realtek: avoid flooding unknown multicast groups"
  * changing function prefixes from rtl83xx_ to rtldsa_
  * renaming MC_PMASK_NO_PORTS_IDX to MC_PMASK_MIN_PORTS_IDX, as it will contain multicast router ports in the next, follow-up patch
  * removing MC_PMASK_NO_PORTS (would be replaced in follow-up patch anyway)
* rtl83xx: forward multicast packets to multicast router ports
  * changing title to "realtek: forward multicast packets to multicast router ports"
  * changing function prefixes from rtl83xx_ to rtldsa_
  * using dev_err() instead of pr_err()
  * renaming MC_PMASK_NO_PORTS_IDX to MC_PMASK_MIN_PORTS_IDX
  * adding "Fixes:" line to commit message

T-X commented:
Changelog v4:
* rebased to current main branch:
  * removed "realtek: fix flooding of unsnoopable multicast addresses", as now already applied via 64ad16993dc423e29ed59effb6d4784e27ca8c31 via #18769

plappermaul reviewed (approved):
Absolutely needed to make this code more readable.

plappermaul commented on target/linux/generic/pending-6.6/696-net-dsa-forward-bridge-switchdev-mcast-router-notifi.patch:
```diff
@@ -0,0 +1,99 @@
+From 60d5de358f5d5745623f3293f7888a1ea47d6d9d Mon Sep 17 00:00:00 2001
```
The other parts (realtek driver adaptions) can stay here.

plappermaul reviewed (changes requested):
Little bit confused with review finalization. So set it now to "request changes".

T-X commented:
Changelog v5:
* rebased to current main branch (no changes)
* realtek: replace more "magic numbers" with macros:
  * using GENMASK() and GENMASK_ULL() for all-ports port masks
  * replacing a ">= 0x9300" with ">= RTL9300_FAMILY_ID", too
* realtek: add port_masks, vlan_profiles and vlans debugfs tables
  * prefixing new functions in the debugfs.c with rtldsa_, too
* realtek: avoid flooding unknown multicast groups
  * replacing a ">= 0x9300" with ">= RTL9300_FAMILY_ID", too
* realtek: only apply multicast snooping with an active querier:
  * adding a few fixes to the Linux bridge patches in sync with what was submitted upstream:
    * removed an "out:" goto label from "net: bridge: mcast: explicitly track active state", which is only used later with "net: bridge: switchdev: notify on mcast active changes"
    * added IFLA_BR_MCAST_ACTIVE_V4 and IFLA_BR_MCAST_ACTIVE_V6 to the br_get_size() check

Also first version for the bridge/switchdev/dsa changes submitted to the upstream Linux kernel here:

* https://patchwork.kernel.org/project/netdevbpf/cover/20250522195952.29265-1-linus.luessing@c0d3.blue/
* openwrt-devel mailing list CC'd: https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043999.html

Neustradamus commented:
@T-X: Good job!

BrainSlayer commented:
root@ECS4125P:~# brctl addbr br0
add bridge failed: No error information

BrainSlayer commented:
i will debug this and will provide a proposal for fix

T-X commented:
> okay. the result is simple. you added to many netlink entries. the limit is 50. and together with other patches which might not be openwrt specific but can be openwrt specific too since i backported a 6.12 feature, your list has a length of 51. so bridge netlink registering fails.

@BrainSlayer thanks for testing and reporting! Yes I ran into this when testing it on Linux master, too, before I submitted the bridge/switchdev/dsa changes upstream. And in there I had bumped that RTNL_MAX_TYPE: https://patchwork.kernel.org/project/netdevbpf/patch/20250522195952.29265-3-linus.luessing@c0d3.blue/

But I think for 6.6 this bump is not needed yet?

BrainSlayer commented:
> > okay. the result is simple. you added to many netlink entries. the limit is 50. and together with other patches which might not be openwrt specific but can be openwrt specific too since i backported a 6.12 feature, your list has a length of 51. so bridge netlink registering fails.
> 
> @BrainSlayer thanks for testing and reporting! Yes I ran into this when testing it on Linux master, too, before I submitted the bridge/switchdev/dsa changes upstream. And in there I had bumped that RTNL_MAX_TYPE: https://patchwork.kernel.org/project/netdevbpf/patch/20250522195952.29265-3-linus.luessing@c0d3.blue/
> 
> But I think for 6.6 this bump is not needed yet?

for me it is, but its not openwrt specific maybe since i'm developing on dd-wrt and just right now on the rtl83x/rtl93xx targets. btw. you should review my latest enhancements since i have now rtl9313 fully working (also sfp ports based on edgecore ecs4125) i also enhanced the dsa driver alot and since you are touching this code right now its maybe wise to take a sneak. i'm in discussion with @plappermaul  to get these changes merged into openwrt. but who knows. 2 more eyes with a look on that matter might help.

plappermaul commented:
@T-X polishing 6.12 kernel with #18935 . Will try to stabilize this week so you prepare this on top.

plappermaul commented:
Waiting to be rebased against 6.12

BrainSlayer commented:
i have ported it to 6.12 but its not my ownership and i cannot steal it. i'm wating this PR to be merged since i have a big patch set which builds on top of it

T-X commented:
@BrainSlayer thanks for having a look at it! will hopefully get back to it this week, to rebase and test it on 6.12. and to split the DSA part out of this PR into a separate one.

Meanwhile I had also made a PR for upstream Linux integration for the DSA/bridge multicast features and had gotten some valuable feedback which I also want to integrate:

https://lwn.net/ml/all/20250522195952.29265-1-linus.luessing@c0d3.blue/
(= https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043999.html)

jonasjelonek committed:
realtek: add v6.18 as testing kernel

Add kernel v6.18 as testing kernel for realtek target.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: add missing symbols for 6.18

Add missing symbols to default config by running make kernel_oldconfig
and using default options for all unanswered symbols.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: phy: 6.18: drop shared priv access helper

Between v6.12 and v6.18, struct phy_package_shared was moved from
include/linux/phylib.h to drivers/net/phy/phy_package.c and thus made
internal/unavailable for other drivers.

Our PHY driver used a helper to access the shared priv. With v6.18, it
doesn't work anymore (due to missing access to internal fields of struct
phy_package_shared) but upstream has an equivalent helper for that. Drop
ours and use the upstream helper.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: pcs: 6.18: fix neg_mode handling

Between 6.12 and 6.18, neg_mode moved from struct phylink_pcs to a
parameter passed to pcs_get_state. Adjust the implementation
accordingly.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: eth: 6.18: fix neg_mode handling in pcs ops

Handling of neg_mode moved from struct phylink_pcs to a parameter passed
to pcs_get_state between 6.12 and 6.18. Adjust the PCS related
definitions in the ethernet driver to fix a compilation issue.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: dsa: 6.18: adjust mac ops and callbacks

Between 6.12 and 6.18, several DSA switch ops have changed.
phylink_mac_* ops were moved out of dsa_switch_ops into dsa_switch using
phylink_mac_ops. Add new definitions for that and adjust the callbacks
accordingly to work with the different signatures.

get_mac_eee() has been dropped and support_eee() was added.

remove_new() has been dropped completely, allowing us to use remove()
again.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

namiltd committed:
kernel: add define for kernel 6.18

Add define for kernel 6.18.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: refresh backport patches

Refreshed patches for generic/backport-6.18 by running
make target/linux/refresh

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: refresh pending patches

Refreshed patches for generic/pending-6.18 by running
make target/linux/refresh

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: refresh hack patches

Refreshed patches for generic/hack-6.18 by running
make target/linux/refresh

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
malta: 6.18: update kernel symbols

Remove symbols no longer present in version 6.18,
add new 6.18 kernel symbols.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
malta: enable 6.18 testing kernel

The malta now supports 6.18 kernel as testing.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

graysky2 committed:
kernel: kmod-fs-nfs-v4: add option for kernel-6.18

Add a config option for the 6.18 series of kernels

Signed-off-by: John Audia <therealgraysky@proton.me>

namiltd committed:
packages: sctp: 6.18: add missing symbol

Add missing CONFIG_SCTP_DEFAULT_COOKIE_HMAC_SHA256 for kernel 6.18

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: update kernel symbols

Remove symbols no longer present in version 6.18,
add new 6.18 kernel symbols.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: net: phy: adapt to kernel 6.18

 - change phy_driver_register and phy_driver_unregister function calls
   to phy_drivers_register and phy_drivers_unregister for compatibility
   with kernel 6.18
 - replace PHY_ANY_ID with "MATCH ANY PHY"

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
package: gpio-button-hotplug: adapt to kernel 6.18

 Change GPIOF_ACTIVE_LOW to gpiod_toggle_active_low

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

jonasjelonek committed:
realtek: phy: 6.18: switch to sfp module caps

Kernel 6.18 uses dedicated sfp_module_caps instead of a plain bitmask
for supported modes. Switch to that the fix the rtl83xx-phy driver.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

namiltd committed:
kbuild: extract modules.builtin.modinfo from vmlinux.o

Partially reverse https://web.git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.18.y&id=39cfd5b12160be4f57df1c3ba60139741c827616
to prevent generation of empty modules.builtin.modinfo and modules.builtin
files that cause modules compilation errors.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
packages: 6.18: update modules files

Update modules files for kernel 6.18.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
mac80211: restore pre-6.14 debugfs_fops handling

Reintroduce old-style debugfs file_operations for kernels < 6.14 and
switch the drivers to use debugfs_short_fops + debugfs_create_file_aux
on >= 6.14. This restores compatibility with older kernels while keeping
the new debugfs API working on 6.14+.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
mac80211: ath9k: gpio: adapt to 6.16+ gpio_chip API change

Linux 6.16 changed the return type of gpio_chip.set() from void to int.
Add a version check so the driver builds correctly on both older and
newer kernels[1].

1. https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.16&id=fb52f3226cab41b94f9e6ac92b1108bce324e700

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: manually rebuild hack patches

Manually rebuild hack patches:
 - 200-tools_portability.patch
 - 204-module_strip.patch
 - 230-openwrt_lzma_options.patch
 - 251-kconfig.patch
 - 259-regmap_dynamic.patch
 - 301-01-mm-permit-to-declare-custom-execmem-alloc-free-funct.patch
 - 301-02-mips-replace-mlong-calls-with-mno-long-calls-if-poss.patch
 - 430-mtk-bmt-support.patch
 - 721-net-add-packet-mangeling.patch
 - 722-net-phy-aquantia-enable-AQR112-and-AQR412.patch
 - 725-net-phy-aquantia-add-PHY_IDs-for-AQR112-variants.patch
 - 750-net-pcs-mtk-lynxi-workaround-2500BaseX-no-an.patch
 - 800-GPIO-add-named-gpio-exports.patch
 - 902-debloat_proc.patch
 - 904-debloat_dma_buf.patch

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
mac80211: fix build failure on kernel 6.18

Kernel since version 6.16 defines `SOCKWQ_ASYNC_NOSPACE` and
SOCKWQ_ASYNC_WAITDATA` in the socket_flags enum, which causes redefinition
errors when compiling the backport package[1].
Since kernel 6.12 also has them defined, the test and definition of these
constants have been completely abandoned. Thanks to this, there are
no compilation errors on either kernel 6.12 or 6.18.

1. https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.16&id=0e81cfd971dc4833c699dcd8924e54a5021bc4e8

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: manually rebuild pending patches

Manually rebuild pending patches:
 - 203-kallsyms_uncompressed.patch
 - 417-mtd-spi-nand-macronix-disable-continuous-read-for-MX.patch
 - 465-m25p80-mx-disable-software-protection.patch
 - 487-mtd-spinand-Add-support-for-Etron-EM73D044VCx.patch
 - 499-mtd-spi-nor-disable-16-bit-sr-for-macronix.patch
 - 530-jffs2_make_lzma_available.patch
 - 681-net-remove-NETIF_F_GSO_FRAGLIST-from-NETIF_F_GSO_SOF.patch
 - 701-netfilter-nf_tables-ignore-EOPNOTSUPP-on-flowtable-d.patch
 - 710-bridge-add-knob-for-filtering-rx-tx-BPDU-pack.patch
 - 732-03-net-ethernet-mtk_eth_soc-optimize-dma-ring-address-i.patch
 - 737-net-ethernet-mtk_eth_soc-add-paths-and-SerDes-modes-.patch
 - 739-03-net-pcs-pcs-mtk-lynxi-add-platform-driver-for-MT7988.patch
 - 739-05-net-pcs-add-driver-for-MediaTek-USXGMII-PCS.patch
 - 801-gpio-gpio-cascade-add-generic-GPIO-cascade.patch
 - 802-OPP-Provide-old-opp-to-config_clks-on-_set_opp.patch
 - 812-PCI-sysfs-enforce-single-creation-of-sysfs-entry-for.patch
 - 850-0023-PCI-aardvark-Make-main-irq_chip-structure-a-static-d.patch

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

graysky2 committed:
mt76: add kernel version compatibility for timer APIs

Add version checks to support both old and new timer APIs across
different kernel versions:

- Use timer_delete_sync() on kernels >= 6.1, fall back to
  del_timer_sync() on older kernels. The timer_delete_sync() function
  was introduced in kernel 6.1 as a replacement for del_timer_sync().

- Use from_timer() for timer callbacks on kernels >= 4.15, fall back
  to container_of() on older kernels. The from_timer() macro was
  introduced in kernel 4.15 as part of the timer API modernization.

- Use hrtimer_setup() on kernels >= 4.15, fall back to manual
  hrtimer_init() and function assignment on older kernels. The
  hrtimer_setup() helper was introduced alongside the timer callback
  changes in kernel 4.15.

- Use timer_container_of() on kernels >= 6.16 and container_of() for
  older kernels. Since the minimum supported kernel is currently 6.12,
  there is no need to add container_of() for older kernels.

Signed-off-by: John Audia <therealgraysky@proton.me>

jonasjelonek committed:
realtek: add patch to extend upstream sfp smbus support

Add patches that extends the upstream smbus support in sfp driver. This
only supports SMBus byte operations and in that case, disables hwmon and
prints warnings because of that. Using the word/block capability of our
I2C controller is totally fine for all that functionality.

The patches are currently pending upstream.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

namiltd committed:
mac80211: fix Module.symvers.tmp generation on kernel 6.18

For kernel 6.18, the generated Module.symvers.tmp file is empty.
This fixes this by copying the contents of the Module.symvers file
directly to Module.symvers.tmp.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: silence unused-variable warnings

This patch silences compiler warnings about unused variables by explicitly
casting them to void where appropriate.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: manually rebuild backport patches

Manually rebuild backport patches:
 - 750-v7.0-net-phy-move-mmd_phy_read-and-mmd_phy_write-to-phyli.patch

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
platform/mikrotik: adjust hc_wlan read-callback for 6.18

Include <linux/version.h> and use preprocessor checks to select the appropriate
prototype and definition for hc_wlan_data_bin_read. The third parameter of bin_attribute
read callbacks was changed to 'const struct bin_attribute *' in Linux 6.18; this
change prevents compilation errors on newer kernels while preserving compatibility
with older ones

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
crypto: eip93: use in-tree driver for kernel 6.18

This patch set introduces support for using the in-tree (mainline)
EIP93 crypto driver for kernel 6.18 and later, along with additional
improvements and fixes:
 - Conditional Kconfig/Makefile handling for crypto-hw-eip93 to enable use of the mainline driver with kernel 6.18+.
 - Patch 926: Use software AES fallback for small requests in the EIP93 driver.
 - Patch 927: Add `mediatek,mtk-eip93` compatible string for upstream kernel device trees.
 - Patch 880-01: Fix kernel panic in driver detach for EIP93 by correcting hash algorithm unregistration.
 - Patch 880-02: Unregister only those crypto algorithms that are actually available in hardware.
 - Patch 880-03: Prevents sleeping in non-sleep crypto paths.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

T-X commented:
If people have a little more patience my suggestion would be:

1) to split out the first few, small cleanup / cosmetic changes into a new PR, to get them merged more quickly
2) mark this / the rest of this PR as Draft
3) get the Linux bridge / DSA changes merged upstream
4) only after 3) remove the Draft again and merge this in OpenWrt

The Draft status would only be to avoid accumulating even more non-upstream Linux patches. Other than that this should work a lot better than the previous v5 version here, has had quite some more testing and fixes now.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -2657,6 +2659,15 @@ rtldsa_mdb_add_ports(struct rtl838x_switch_priv *priv, u64 mac, u16 vid,
+
+	return rtldsa_mdb_add_ports(priv, mac, vid, MC_PMASK_ALL_PORTS_IDX, -1);
+}
+
```
Extra info here: This should have been "priv->family_id". But we do not want that.

T-X replied:
Decided on adding this new commit 49c58891a93f4e4bbfe9b147bde4a1063d48b85a now. Doesn't feel great, but I don't have an rtl93xx device to properly test the other approach, that is to just remove this priv->id check in rtl83xx_port_mdb_add() while leaving the .port_mdb_* handlers in place, at the moment.

namiltd committed:
serial: 8250: 6.18: add UPIO_AU case to set_io_from_upio()

This patch fixes the "Unsupported UART type 4" error by adding support
for the UPIO_AU I/O type in set_io_from_upio() (used by Au1x00 and
RT288x). It assigns the correct serial input/output functions
(io_serial_in, io_serial_out) for proper operation when UPIO_AU is
configured.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

jonasjelonek committed:
realtek: 6.18: remove upstreamed patches

Remove patches which are accepted upstream and were backported to
pre-6.18.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: 6.18: adapt SFP SMBus patches

Adapt our downstream patches for SMBus SFP functionality to kernel 6.18.
Upstream gained support for SMBus byte access on SFP and a patch to
extend that to word and block access is pending.

These downstream patches aren't acceptable for upstream and will need to
be adapted. However, keep them for now to maintain functionality.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
realtek: 6.18: refresh and adapt remaining patches

Refresh and adapt the remaining patches as needed.

314-irqchip-irq-realtek-rtl-add-VPE-support.patch needed manual
adoption, all others were just refreshed.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

namiltd committed:
kernel/generic: create files for v6.18 (from v6.12)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
kernel/generic: restore files for v6.12

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
generic: 6.18: remove obsolete backport patches

Remove obsolete backport patches already included in kernel 6.18.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
kernel/malta: create files for v6.18 (from v6.12)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
kernel/malta: restore files for v6.12

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

jonasjelonek committed:
kernel/realtek: create files for v6.18 (from v6.12)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek committed:
kernel/realtek: restore files for v6.12

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

T-X committed:
realtek: rtl93xx: remove MDB handlers instead of checking id

Right now the MDB code part should be unused on rtl93xx as after a
priv->id check we return -EOPNOTSUPP. With unset MDB handlers /
handlers set to NULL this should have the same result. And let's
us avoid this priv->id check in the code.

ToDo: Once someone can verify the MDB code on an rtl93xx device then we
can add these handlers in again.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: remove currently unused VLAN profile_id 1

So far we are only associating and using a VLAN profile ID of 0 to any
VLAN ID, VLAN profile ID 1 is unused. Therefore we can safely remove its
initialization, just like we don't initialize any of VLAN profile IDs
2 to 7 on rtl83xx/rtl930x or 2 to 15 on rtl931x.

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>

T-X committed:
realtek: prepare VLAN profile IDs for IPv4/v6 active/inactive states

In preparation for the new multicast active tracking setup a VLAN
profile ID for each combination of IPv4 or IPv6 in active or inactive
state.

This allows us to have individual per protocol family flooding
behaviour, depending on the presence or absence of an IGMP and/or MLD
querier, for instance. Or per protocol family multicast router ports.

For instance if there is no IGMP querier for IPv4 for a set of VIDs
we can flood all IPv4 multicast traffic on ports X, Y and Z. While at the
same time with an MLD querier present and a multicast router port on Z
we can at the same time flood all IPv6 multicast traffic only on port Z.

Note: We unfortunately have a very limited number of VLAN profile IDs,
8 on rtl83xx/rtl930x and 16 on rtl931x. So we can't set invidual flooding
behaviour for each and any of the potential 4095 VLAN IDs. Which can
lead to unnecessarily forwarding of a multicast packet on one VID to a
port which is a actually only a multicast router port for another VID.
(ToDo: But we could dynamically allocate and set up per VID VLAN profile
IDs if we use only a few VIDs. Until we run out of the small VLAN profile
ID space. Then we could / would have to switch to VLAN profile IDs shared
between VIDs again.)

Signed-off-by: Linus Lüssing <linus.luessing@c0d3.blue>
